Introduce Trust Book in Book Viewlet

joberume opened the pull request:
This change introduces a "Trust Book" function that trusts all notebooks in a Book. If a new notebook is introduced after having trusted the book, and it is not added to the table of contents of that book, then that trust will not be extended to that notebook (no inadvertent permission granting). 

This PR fixes #8864

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -7,17 +7,21 @@ import * as azdata from 'azdata';
 import * as loc from '../common/localizedConstants';
 
+const localize = nls.loadMessageBundle();
+const msgBookTrusted = localize('msgBookTrusted', "Book is now automatically trusted.");
```
It's preferable to put all strings in the https://github.com/microsoft/azuredatastudio/blob/master/extensions/notebook/src/common/localizedConstants.ts file

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -57,6 +64,16 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 		return this._initializeDeferred.promise;
 	}
 
+	trustBook(bookTreeItem: BookTreeItem): any {
```
Return should be voice instead of any

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -57,6 +64,16 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
+		let hasTrustedBook = this._bookTrustManager.setBookAsTrusted(bookTreeItem);
+
+		if (hasTrustedBook) {
+			vscode.window.showInformationMessage(msgBookTrusted);
```
All vscode/azdata API calls should go through https://github.com/microsoft/azuredatastudio/blob/master/extensions/notebook/src/common/apiWrapper.ts (I see this file isn't doing that currently, but for new stuff let's make sure we're doing that)

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -112,8 +129,18 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
+				let normalizedResource = path.normalize(resource);
+				if (!cache[normalizedResource] && this._bookTrustManager.isNotebookTrustedByDefault(normalizedResource)) {
+					let openDocumentListenerUnsubscriber = azdata.nb.onDidOpenNotebookDocument(function (document: azdata.nb.NotebookDocument) {
+						document.setTrusted(true);
```
How much overhead would there be to just always setTrusted to true if it's trusted by default? (suggesting get rid of the cache here)

joberume replied:
Not sure how often this would happen, but users are allowed to "untrust" a notebook after having opened it. Remembering which we already trusted by default will stop us from trusting it again and overriding the user's last state.

Charles-Gagnon replied:
But this is only called when the document opens - and closing a notebook clears the trusted status anyways

joberume replied:
Looks like closing the notebook does not clear the trust. There seems to be a cache in the notebook itself:
![cached_trust](https://user-images.githubusercontent.com/52225468/75810272-25fecd80-5d3f-11ea-89f3-9e2bd4c9835f.gif)

Charles-Gagnon replied:
Ah, yeah I didn't try opening from a book. So that seems fine then. Although I'd run this by @chlafreniere / @yualan to see if that behavior makes sense.

I'd suggest adding a comment explaining why the cache is necessary though if you do keep this logic.

chlafreniere replied:
The cache isn't in the notebook. We use mementos. Take a look at src/sql/workbench/services/notebook/browser/notebookServiceImpl.ts

Charles-Gagnon commented on src/sql/azdata.d.ts:
```diff
@@ -4314,6 +4314,11 @@ declare module 'azdata' {
+			/**
+			 * Sets the notebook to trusted by default.
+			 */
+			setTrusted(state: boolean);
```
This should go into https://github.com/microsoft/azuredatastudio/blob/master/src/sql/azdata.proposed.d.ts

(just define NotebookDocument in there - the two definitions are combined during build time)

Charles-Gagnon commented on src/sql/workbench/api/browser/mainThreadNotebookDocumentsAndEditors.ts:
```diff
@@ -362,6 +362,17 @@ export class MainThreadNotebookDocumentsAndEditors extends Disposable implements
 		return Promise.resolve(this.doOpenEditor(resource, options));
 	}
 
+	$trySetTrusted(uri: UriComponents, isTrusted: boolean = true): Promise<boolean> {
```
I'd get rid of the default value for isTrusted - this should always be explicitly specified

Charles-Gagnon commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,89 @@
+
+	getNotebookWorkspaceFolder(notebookUri: string): vscode.WorkspaceFolder {
+		let workspace = vscode.workspace;
+		let workspaceFolder = workspace.workspaceFolders.find(wsFolder => {
```
I don't believe books are always going to be opened as a workspace folder. Users can open books directly which just adds it to the tree view but doesn't actually open it in the workspace.

Charles-Gagnon commented:
You also have a merge conflict so make sure you're merging in the latest changes from master

Charles-Gagnon commented on extensions/notebook/package.json:
```diff
@@ -282,6 +299,10 @@
+        },
+        {
+          "command": "notebook.command.trustBook",
+          "when": "false"
```
You don't want to have this be available in the command palette?

Charles-Gagnon replied:
Why not? That seems like a reasonable thing to have

joberume replied:
No reason in particular. Have made the change.

Charles-Gagnon commented on extensions/notebook/package.nls.json:
```diff
@@ -6,6 +6,7 @@
 	"notebook.useExistingPython.description": "Local path to a preexisting python installation used by Notebooks.",
 	"notebook.overrideEditorTheming.description": "Override editor default settings in the Notebook editor. Settings include background color, current line color and border",
 	"notebook.maxTableRows.description": "Maximum number of rows returned per table in the Notebook editor",
+	"notebook.trustedBooks.description": "Books to trust automatically.",
```
Hm, I'd consider trying to make it clearer what this is doing. Something that includes mentioning that it will result in child notebooks being trusted (the book itself doesn't really have a concept of trust)

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -38,7 +42,8 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 		this.prompter = new CodeAdapter();
-
+		this._bookTrustManager = new BookTrustManager(this.books);
+		this._apiWrapper = new ApiWrapper();
```
Don't create a new ApiWrapper - use the one in the AppContext created in extension.ts

Charles-Gagnon commented on src/sql/azdata.proposed.d.ts:
```diff
@@ -29,6 +29,12 @@ declare module 'azdata' {
 		export function getConnection(uri: string): Thenable<ConnectionProfile>;
 	}
 
+	export interface NotebookDocument {
```
This needs to be in the nb namespace as well (that's why you were getting the error about setTrusted not existing on the type). Once you do that you should be able to get rid of the warning suppression you put there

Charles-Gagnon commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,146 @@
+		let normalizedContainerPath: string = workspace ? path.normalize(workspace?.uri.fsPath) : undefined;
+
+		if (!normalizedContainerPath) {
+			normalizedContainerPath = this.books
```
So if I'm understanding this correctly this is where you'll look through all open books if you didn't find it in the open workspaces? A couple questions :

1. Why wouldn't you just look in the open books directly? That should contain all books right? (IIRC we load all books in a workspace when that workspace is opened)
2. Did you look and see if there's a helper of some sort to see if a book is currently open? (basically this logic but in a more centralized place). If not I would suggest adding something like that since this seems like a general case that we may want in the future

joberume replied:
Ah yes, you make a very good point. Initially the design was to trust the directory but it has shifted to the notebooks in the table of contents. I have shifted the design to leverage the books instead of directories. Overall, I think it's more readable. Thanks for pointing it out.

Charles-Gagnon commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,146 @@
+
+	private static notebookConfiguration: string = 'notebook';
+	private static notebookTrustedBooksConfiguration: string = 'trustedBooks';
+	private trustedLocalBooks: Record<string, boolean> = {};
```
I think using a dictionary makes more sense here - having the entries be properties is kind of weird

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -26,24 +26,24 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 	public currentBook: BookModel;
 
-	constructor(workspaceFolders: vscode.WorkspaceFolder[], extensionContext: vscode.ExtensionContext, openAsUntitled: boolean, view: string) {
+	constructor(workspaceFolders: vscode.WorkspaceFolder[], extensionContext: vscode.ExtensionContext, openAsUntitled: boolean, view: string, apiWrapper?: ApiWrapper) {
```
Any reason you have this be optional? The apiWrapper should never be undefined - we can guarantee that

Charles-Gagnon commented on extensions/notebook/package.json:
```diff
@@ -154,6 +162,15 @@
         }
       },
+      {
+        "command": "notebook.command.trustBook",
```
Also - there currently doesn't appear to be a way to untrust a book (without modifying the config manually)? Is that by design?

joberume replied:
That's correct. The design so far does not include providing an "Untrust" functionality, Let me forward you the email.

joberume replied:
This is a good point. I raised the same point with the designer in the email that I sent you.

chlafreniere replied:
I actually like that there isn't a way to untrust a book. I think it's a much less used scenario. But that's just me 😄

Charles-Gagnon replied:
If @chlafreniere doesn't think there's a need for untrusting a book then I'd go with that. Sync up with him on this.

joberume replied:
Synced up with Chris. Will be removing it.

chlafreniere commented on src/sql/workbench/api/common/extHostNotebookDocumentData.ts:
```diff
@@ -50,6 +49,9 @@ export class ExtHostNotebookDocumentData implements IDisposable {
 				save() { return data._save(); },
+				// ignore error "setTrusted' does not exist in type 'NotebookDocument'." as it is merged from azdata.proposed.d.ts
+				// @ts-ignore
+				setTrusted(isTrusted) { data._setTrusted(isTrusted); },
```
@Charles-Gagnon I thought we merged azdata and the proposed version at compile time? Am I misremembering something?

joberume replied:
Correct. Already moved it to proposed, but forgot to remove the comment. Thanks!

chlafreniere commented on extensions/notebook/src/common/localizedConstants.ts:
```diff
@@ -21,8 +21,10 @@ export const confirmReplace = localize('confirmReplace', "Folder already exists.
 export const openMarkdownCommand = localize('openMarkdownCommand', "Open Markdown");
 export const openExternalLinkCommand = localize('openExternalLinkCommand', "Open External Link");
-
+export const msgBookTrusted = localize('msgBookTrusted', "Book is now automatically trusted.");
```
Can we change this string from "automatically" to something that includes "workspace"

chlafreniere commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,137 @@
+
+export class BookTrustManager implements IBookTrustManager {
+
+	private static notebookConfiguration: string = 'notebook';
```
we can use notebookConfigKey (constants.ts) here instead of defining it again.

joberume replied:
Thanks for the suggestion. Made the change :)

chlafreniere commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,137 @@
+export class BookTrustManager implements IBookTrustManager {
+
+	private static notebookConfiguration: string = 'notebook';
+	private static notebookTrustedBooksConfiguration: string = 'trustedBooks';
```
let's put this in constants.ts so that this is all in one spot.

chlafreniere commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,137 @@
+/*---------------------------------------------------------------------------------------------
```
this class looks quite test-able. Maybe during committer time tests could be added 😄?

joberume replied:
I am a little stuck with the Debugging Extension Unit Tests hanging indefinitely. Need to follow up with team members on Thursday to see if I can get unblocked.

Charles-Gagnon replied:
You can also do print (console) debugging too - that's what I usually fall back on when I start having issues debugging (I have this kind of thing happen occasionally too for the extension tests, haven't had time to look into it much yet unfortunately)

joberume replied:
Gotcha. I managed to run the tests by getting rid of the rest of the modules in the `scripts\test-extensions-unit.bat` script.

chlafreniere commented on src/sql/azdata.proposed.d.ts:
```diff
@@ -29,6 +29,14 @@ declare module 'azdata' {
+	export namespace nb {
+		export interface NotebookDocument {
+			/**
+			 * Sets the notebook to trusted by default.
```
Please update the comment. Probably shouldn't include the word "default".

chlafreniere commented on src/sql/workbench/api/browser/mainThreadNotebookDocumentsAndEditors.ts:
```diff
@@ -362,6 +362,17 @@ export class MainThreadNotebookDocumentsAndEditors extends Disposable implements
 		return Promise.resolve(this.doOpenEditor(resource, options));
 	}
 
+	$trySetTrusted(uri: UriComponents, isTrusted: boolean): Promise<boolean> {
```
Is there a reason why we can't leverage the notebookservice here instead of adding another API? Could we leverage the serializeNotebookStateChange() method (or refactor that method to also include another "updateTrust" method)?

joberume replied:
Do you mean extend the `NotebookService` to the extension via an RPC call? API is updated to invoke it from the extension. `NotebookService` is hosted in the main code base/workbench. Maybe I am missing something here

chlafreniere commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,137 @@
+	}
+
+	getTrustedBookPathsInConfig(): string[] {
+		let trustedBookPaths: string[] = this.trustedLocalBooks;
```
Seems like we have a couple of scenarios here:
- Book is part of a workspace. What happens if a user opens the book in ADS outside of the workspace? Should it be trusted? I'm asking because I don't know the right answer 😄 
- Book isn't part of a workspace. This is the simpler scenario because it should always be trusted.

Can we leverage the same mementos mechanism somehow that we use for single notebooks? If you're not familiar with it, take a look at notebookServiceImpl.

joberume replied:
1. Currently, the way that it works now is that a separate book will be trusted, regardless if it is part of the workspace or not.
2. So if a book is not part of a workspace it is automatically trusted? What if I copied someone's book from a share into my drive? What constitutes the safety of the notebook? Currently I do not make this destinction in this change. 

Memento comment below

Charles-Gagnon replied:
I'm not sure if @chlafreniere is suggesting to use the actual same trusted notebook cache - but if the suggestion is just to use mementos in general that's available to extensions through extensionContext.globalState. 

Example : https://github.com/microsoft/azuredatastudio/blob/master/extensions/cms/src/extension.ts#L20

joberume replied:
Ah yes, I found this out after posting this comment 😁. Thanks for the pointer it helped out!
Leveraged the Mementos 😄

joberume commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -7,38 +7,43 @@ import * as azdata from 'azdata';
 	private _initializeDeferred: Deferred<void> = new Deferred<void>();
-
 	private _openAsUntitled: boolean;
+	private _trustResourceCache: Record<string, boolean> = {};
```
Cache was aliased with name `cache`. Not needed, switched back to original name.

joberume commented:
> memento

The memento framework is situated in the workbench contribution, whereas this change is in the extension. I can import the library over, but it seems that it's tied up well with the rest of the code base due to the dependencies. Were you thinking of a RPC API to link the states?

joberume commented:
> Couple of things that I left comments on. Would be great to understand if we can use mementos.

The memento framework is situated in the workbench contribution, whereas this change is in the extension. I can import the library over, but it seems that it's tied up well with the rest of the code base due to the dependencies. Were you thinking of a RPC API to link the states?

joberume commented:
> > Couple of things that I left comments on. Would be great to understand if we can use mementos.
> 
> The memento framework is situated in the workbench contribution, whereas this change is in the extension. I can import the library over, but it seems that it's tied up well with the rest of the code base due to the dependencies. Were you thinking of a RPC API to link the states?

@chlafreniere

joberume commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,89 @@
+		let config: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration(BookTrustManager.notebookConfiguration);
+		let trustedBookDirectories: string[] = config.get(BookTrustManager.notebookTrustedBooksConfiguration);
+		let notebookUriWithoutBase: string = notebookUri.substring(path.normalize(workspace.uri.fsPath).length + 1);
+		return trustedBookDirectories.find(dir => notebookUriWithoutBase.startsWith(dir));
```
Latest change handles this better

Charles-Gagnon commented on extensions/notebook/package.json:
```diff
@@ -282,6 +304,12 @@
           "when": "false"
+        },
+        {
+          "command": "notebook.command.trustBook"
```
This needs to have a when clause to only show when on a book tree item :

`"when": "view == bookTreeView && viewItem == savedBook`

joberume replied:
Nice. Thanks!

Charles-Gagnon commented on extensions/notebook/package.nls.json:
```diff
@@ -6,6 +6,7 @@
 	"notebook.useExistingPython.description": "Local path to a preexisting python installation used by Notebooks.",
 	"notebook.overrideEditorTheming.description": "Override editor default settings in the Notebook editor. Settings include background color, current line color and border",
 	"notebook.maxTableRows.description": "Maximum number of rows returned per table in the Notebook editor",
+	"notebook.trustedBooks.description": "Notebooks contained by these books will automatically be trusted.",
```
`contained in`

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -7,38 +7,43 @@ import * as azdata from 'azdata';
 		this.prompter = new CodeAdapter();
-
+		this._bookTrustManager = new BookTrustManager(this.books);
+		this._apiWrapper = apiWrapper;
```
Just declare this in the constructor as `private _apiWrapper: ApiWrapper` and you can get rid of the extra declaration.

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -113,6 +158,19 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
+				// let us keep a list of already visited notebooks so that we do not trust them again, potentially
+				// overriding user changes
+				let normalizedResource = path.normalize(resource);
+				let self = this;
```
You shouldn't need to use self here if you use a lambda instead of function declaration 

`(document: azdata.nb.NotebookDocument) => { ... }`

Charles-Gagnon commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,161 @@
+		if (!workspaceDetails) {
+			this.workspaceDetails = {
+				get getConfiguration() {
+					return vscode.workspace.getConfiguration;
```
This should also be using the ApiWrapper

joberume replied:
Updated to use ApiWrapper

Charles-Gagnon commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,161 @@
+					return vscode.workspace.getConfiguration;
+				},
+				get workspaceFolders() {
+					return [].concat(vscode.workspace.workspaceFolders || []);
```
Why not just `return vscode.workspace.workspaceFolders || []`?

(this should also use ApiWrapper)

joberume replied:
`vscode.workspace.workspaceFolders` is const and typescript complains. Switching this to use ApiWrapper could help with this.

joberume replied:
# resolved

Charles-Gagnon commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,161 @@
+	private static notebookTrustedBooksConfiguration: string = constants.trustedBooksConfigKey;
+	private trustedLocalBooks: string[] = [];
+
+	constructor(private books: BookModel[], private workspaceDetails?: IBookTrustManagerWorkspaceDetails) {
```
I'm not really a fan of how we're passing in the list of books and then just assuming that that specific array is going to always have the latest set. There should really be a BookManagerService or something that does this that things like this can query for the current set of books and whatnot. I know this is adding onto an already largish set of changes but this is going to start getting messy fast if we keep adding things like this that do their own thing.

@chlafreniere @MaddyDev Thoughts on this?

joberume replied:
# resolved

Charles-Gagnon commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,161 @@
+		if (trustedBookItem) {
+			let fullBookBaseUriWithContent = path.join(trustedBookItem.book.root, 'content', path.sep);
+			let requestingNotebookFormattedUri = normalizedNotebookUri.replace(fullBookBaseUriWithContent, '').replace('.ipynb', '');
+			let notebookInTOC = trustedBookItem.book.tableOfContents.sections.find(jupyterSection => {
```
This kind of logic (determining if a particular Notebook is part of a book) feels like it really needs to be pulled into a more generic place like the BookModel

joberume replied:
Moved it out to `BookTreeItem`. For me it does not make sense to put it in the `BookModel` as it seems to more appropriately represent a BookModelNavigator/View. This entity has multiple books and so no unique identifier for books (perhaps maybe by root, but that feels more like a pseudo identifier) can be supplied to specify the search. Ultimately, if we wanted in the `BoodModel` we could aggregate if from the individual `BookTreeItem`s we have here.

Charles-Gagnon commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,161 @@
+		let trustablePaths: string[];
+		let bookPathsInConfig: string[] = this.getTrustedBookPathsInConfig();
+
+		if (this.hasWorkspaceFolders()) {
```
So why do you need to check workspace folders here anyways? I thought that we always added books contained in a workspace to the list of books when we opened that workspace - is this not the case?

joberume replied:
Not exactly. A workspace can contain multiple books. Trusting one book in the workspaces does not extend that same trust to other books in the same workspace. The `getTrustedBookPathsInConfig` retrieves the trusted books in a particular workspace. In our team, we have 4 books in one workspace.

Charles-Gagnon replied:
Yeah - so what I'm saying is that we already parse the workspaces for existing books when initializing the tree : https://github.com/microsoft/azuredatastudio/blob/master/extensions/notebook/src/book/bookTreeView.ts#L50

I don't think this trust logic needs to know about workspaces at all. It should just keep a list of "trusted books" and then when a book is opened if it's in the list of trusted books it goes and trusts all the notebooks it contains.

Does that make sense?

joberume replied:
Ah yes, I see what you are saying. The aim here was to store in the workspace configuration, the paths of the books, relative to the workspace, so that trusting relationship can survive relocation. This is specifically needed in our scenario in that we want users to save the trust relationship on file, and share it with other team members. For example, the workspace configuration may look like this:
```
{
	"folders": [
		{
			"path": "."
		}
	],
	"settings": {
		"notebook.trustedBooks": [
			"Books/BookA",
			"Books/BookB"
		]
	}
}
```
Where the `path` property is all the files where the workspace file lives. The model, currently does not have a unique identifier other than the root folder, but that's an absolute path. To opt to not store full paths in configuration, workspace info is pulled into the trust manager to workout the relative paths so that it is cleaner. So if the workspace folder is located at `C:\temp\workspace` (books would be located at `C:\temp\workspace\Books\BookA` and `C:\temp\workspace\Books\BookB`), then saving it as a relative path allows us to maintain the trust relationship, say for when the workpace file is relocated to `D:\newfolder\`.

You are right, if we were to only to make a book trust on a user to user basis, but for extending this trust for other people, the above helps. 

Perhaps there is another avenue for expressing this in the code base. Approached it this by encapsulating the logic in this module for now, and refactor later as the need for it became broader. 

Let me know what you think.

Charles-Gagnon replied:
Ah, got it. Yeah that's an interesting use case. Definitely behavior I think we'll want although I'm not sure this is the best way. Let me think on it and we can sync up next week - sound good?

Charles-Gagnon replied:
Ok - we can discuss this more later but I looked at this and generally think what you're doing is fine. But I did notice that it looks like this code is set up to assume that all trusted books are in workspaces. I think this should be expanded to allow trusting a book that was opened directly as well. We'll need to discuss how we'd want to determine whether a book should be saved as a relative path or an absolute path though.

joberume replied:
Synced in our meeting. Workspace leverage relative path, folders absolute path.

joberume replied:
# resolved

Charles-Gagnon commented on extensions/notebook/src/test/book/bookTrustManager.test.ts:
```diff
@@ -0,0 +1,337 @@
+			trustedSubFolders = ['/SubFolder/'];
+
+			workspaceDetails = {
+				// @ts-ignore
```
Why the ts-ignore here?

joberume replied:
Remnant comment

Charles-Gagnon commented on extensions/notebook/src/test/book/bookTrustManager.test.ts:
```diff
@@ -0,0 +1,337 @@
+		let workspaceDetails: object;
+		let books: any[];
+
+		this.beforeEach(() => {
```
You shouldn't need the this here

joberume replied:
#resolved

Charles-Gagnon commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,161 @@
+		let allBooks: BookTreeItem[] = this.getAllBooks();
+		let trustableBookPaths = this.getTrustableBookPaths();
+		let trustableBooks: BookTreeItem[] = allBooks
+			.filter(bookItem => trustableBookPaths.some(trustableBookPath => trustableBookPath === path.join(bookItem.book.root, path.sep)));
```
I keep seeing logic around adding and removing path separators which I'm a bit worried about - @MaddyDev already had to do some custom work around that to make the paths correctly open in editors and I feel like there's assumptions being made here that might not always be true. For example you use path.sep here, but is that always going to match what book.root has for its separators? The thing Maddy did involved switching to using posix paths so would always be /, even on Windows machines.

Have you tested this code on both mac/linux and windows?

Charles-Gagnon replied:
So yeah - you're getting some test failures on the linux build which seems to imply my concerns are valid here. Let's make sure to fully test this stuff on both platforms to verify it's working correctly.

joberume replied:
Oh I see, I dismissed it because I thought it was this build:
![image](https://user-images.githubusercontent.com/52225468/76120031-79725500-5fa5-11ea-9b0c-881d3f163218.png)

For some reason I thought the PR validation check was getting sign offs for the PR. Will look into this.

joberume replied:
#resolved

MaddyDev commented on extensions/notebook/src/book/bookTrustManager.ts:
```diff
@@ -0,0 +1,161 @@
+		return isNotebookTrusted;
+	}
+
+	isNotebookPartOfBooks(notebookUri: string, books: BookTreeItem[]) {
```
>isNotebookPartOfBooks [](start = 1, length = 21)

When you do have access to BookModel, it has get all notebooks method, which gives you the list in the toc. Why not check directly if the path exists on the list?

joberume replied:
Moved logic out to `BookTreeItem`. See above comment. #resolved

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -57,6 +60,46 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
+		}
+	}
+
+	untrustBook(bookTreeItem?: BookTreeItem): void {
```
There's still some untrusting code here

joberume replied:
Got it. Had left it there intentionally as an API provision, but yes, not really used. Will remove this from here, but will keep it in the trustmanager but left a method in the trust manager since that one is used by the tests.

Charles-Gagnon commented on extensions/notebook/src/book/bookTreeItem.ts:
```diff
@@ -161,6 +161,24 @@ export class BookTreeItem extends vscode.TreeItem {
+	 * Determines whether a notebook is part of this book.
+	 * @param notebookUri Notebook uri Full URI of the notebook.
+	 */
+	public hasNotebook(notebookUri: string): boolean {
```
I don't think I agree that this is the best place for this. The tree item should be just about the actual tree itself and not necessarily know anything about the contents of the book (only that it may have child book tree items)

Also we already have a getNotebook method on that model. Why can't you just use that? [https://github.com/microsoft/azuredatastudio/blob/master/extensions/notebook/src/book/bookModel.ts#L50](https://github.com/microsoft/azuredatastudio/blob/master/extensions/notebook/src/book/bookModel.ts#L50)

joberume replied:
# Resolved

Charles-Gagnon commented on src/sql/workbench/services/notebook/browser/notebookServiceImpl.ts:
```diff
@@ -606,4 +606,22 @@ export class NotebookService extends Disposable implements INotebookService {
+	 * @param isTrusted True if the notebook is to be trusted, false otherwise.
+	 */
+	async setTrusted(notebookUri: URI, isTrusted: boolean): Promise<boolean> {
+		if (isTrusted) {
```
I'd rather we get rid of all untrusted logic if we're not going to support that. You mentioned that you need it for testing - could you point out where that is? There should be some other way to achieve that functionality (such as clearing the config or something) without needing to use this.

joberume replied:
# resolved

joberume commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -60,6 +64,30 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
+		if (bookPathToTrust) {
+			let trustChanged = this._bookTrustManager.setBookAsTrusted(bookPathToTrust);
+
+			if (this._apiWrapper) {
```
# resolved

joberume commented on extensions/notebook/src/book/bookTreeView.ts:
```diff
@@ -41,6 +44,7 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 		this.viewId = view;
 		this.prompter = new CodeAdapter();
+		this._bookTrustManager = new BookTrustManager(this.books, apiWrapper);
 		this._apiWrapper = apiWrapper ? apiWrapper : new ApiWrapper();
```
# resolved

joberume commented:
> Some issues I'm seeing :
> 
> 1. The trust isn't set immediately upon opening and when it is set it doesn't update the toolbar so it still shows untrusted. If you close and reopen the notebook it'll show correctly though
> 2. There seems to be a lot of issues around getting the trust to persist correctly. I'm seeing a lot of instances where I've trusted a book but the notebooks under it are still showing as untrusted.
> 
> So those are obviously big issues. But at this point I think that aside from a few more clean up items I think we should just get this merged in and then your team can start fully testing it and fixing things one by one. That'll make it a lot easier to track and review the changes. This PR is too overloaded at this point and it's getting hard to review and track all the issues.
> 
> @chlafreniere You good with that?

Investigated this. The path for updating the trust notebook state via Mementos only reflected changes after the notebook is opened. Fixed this by updating the live notebook as well.

chlafreniere commented:
> > Some issues I'm seeing :
> > 
> > 1. The trust isn't set immediately upon opening and when it is set it doesn't update the toolbar so it still shows untrusted. If you close and reopen the notebook it'll show correctly though
> > 2. There seems to be a lot of issues around getting the trust to persist correctly. I'm seeing a lot of instances where I've trusted a book but the notebooks under it are still showing as untrusted.
> > 
> > So those are obviously big issues. But at this point I think that aside from a few more clean up items I think we should just get this merged in and then your team can start fully testing it and fixing things one by one. That'll make it a lot easier to track and review the changes. This PR is too overloaded at this point and it's getting hard to review and track all the issues.
> > @chlafreniere You good with that?
> 
> Investigated this. The path for updating the trust notebook state via Mementos only reflected changes after the notebook is opened. Fixed this by updating the live notebook as well.

Can you clarify what the remaining known issues are?

chlafreniere commented:
Please be sure to merge latest from master; then the component detection issues will go away and builds will succeed.

joberume commented:
> > Investigated this. The path for updating the trust notebook state via Mementos only reflected changes after the notebook is opened. Fixed this by updating the live notebook as well.



> > > Some issues I'm seeing :
> > > 
> > > 1. The trust isn't set immediately upon opening and when it is set it doesn't update the toolbar so it still shows untrusted. If you close and reopen the notebook it'll show correctly though
> > > 2. There seems to be a lot of issues around getting the trust to persist correctly. I'm seeing a lot of instances where I've trusted a book but the notebooks under it are still showing as untrusted.
> > > 
> > > So those are obviously big issues. But at this point I think that aside from a few more clean up items I think we should just get this merged in and then your team can start fully testing it and fixing things one by one. That'll make it a lot easier to track and review the changes. This PR is too overloaded at this point and it's getting hard to review and track all the issues.
> > > @chlafreniere You good with that?
> > 
> > 
> > Investigated this. The path for updating the trust notebook state via Mementos only reflected changes after the notebook is opened. Fixed this by updating the live notebook as well.
> 
> Can you clarify what the remaining known issues are?

I believe issue #1 and #2 are similar. Was not able to repro this. Need to talk to get Book.
